roadmap/implementers-guide/src/runtime-api/README.md

```diff
@@ -0,0 +1,172 @@
+
+Although Runtime APIs are often used for simple storage access, they are actually empowered to do arbitrary computation. The implementation of the Runtime APIs lives within the Runtime as Wasm code and exposes extern functions that can be invoked with arguments and have a return value. Runtime APIs have access to a variety of host functions, which are contextual functions provided by the Wasm execution context, that allow it to carry out many different types of behaviors.
+
+Abilities provided by host functions includes:
+  * State Access
+  * Offchain-DB Access
+  * Submitting transactions to the transaction queue
+  * Optimized versions of cryptographic functions
+  * More
+
+So it is clear that Runtime APIs are a versatile and powerful tool to leverage the state of the chain. In general, we will use Runtime APIs for these purposes:
+  * Access of a storage item
+  * Access of a bundle of related storage items
+  * Deriving a value from storage based on arguments
+  * Submitting misbehavior reports
+
+More broadly, we have the goal of using Runtime APIs to write Node-side code that fulfills the requirements set by the Runtime. In particular, the constraints set forth by the [Scheduler](../runtime/scheduler.md) and [Inclusion](../runtime/inclusion.md) modules. These modules are responsible for advancing paras with a two-phase protocol where validators are first chosen to validate and back a candidate and then required to ensure availability of referenced data. In the second phase, validators are meant to attest to those para-candidates that they have their availability chunk for. As the Node-side code needs to generate the inputs into these two phases, the runtime API needs to transmit information from the runtime that is aware of the Availability Cores model instantiated by the Scheduler and Inclusion modules.
+
+Node-side code is also responsible for detecting and reporting misbehavior performed by other validators, and the set of Runtime APIs needs to provide methods for observing live disputes and submitting reports as transactions.
+
+The next sections will contain information on specific runtime APIs.
+
+## Validators
+
+Yields the validator-set at the state of a given block. This validator set is always the one responsible for backing parachains in the child of the provided block.
+
+```rust
+fn validators() -> Vec<ValidatorId>;
+```
+
+## Validator Groups
+
+Yields the validator groups used during the current session. The validators in the groups are referred to by their index into the validator-set.
+
+```rust
+/// A helper data-type for tracking validator-group rotations.
+struct GroupRotationInfo {
+	session_start_block: BlockNumber,
+	group_rotation_frequency: BlockNumber,
+	now: BlockNumber,
+}
+
+impl GroupRotationInfo {
+	/// Returns the index of the group needed to validate the core at the given index,
+	/// assuming the given amount of cores/groups.
+	fn group_for_core(core_index: usize, cores: usize) -> usize;
```

Comment:
```suggestion
	fn group_for_core(core_index: ValidatorIndex, core: usize) -> usize;
```

If we have a `ValidatorIndex` newtype, we may as well use it.

Reply:
But these aren't validator indices, they're core indices. We do have a `CoreIndex` type, although I would like to keep it private to the scheduler module. It drifted in #1312 ; after #1411 we can move them back a bit.